core/libraries/batch/JobHandlers/PreviewEventDeletion.php

```diff
@@ -112,6 +136,51 @@ protected function createModelObjNodes($model_objs, $dont_traverse_models = [])
         return $nodes;
     }
 
+    /**
+     * Gets all the transactions related to these events that aren't related to other events. They'll be deleted too.
+     * (Ones that are related to other events can stay around until those other events are deleted too.)
+     * @since $VID:$
+     * @param $event_ids
+     * @return array of transaction IDs
+     */
+    protected function getTransactionsToDelete($event_ids)
+    {
+        if (empty($event_ids)) {
+            return [];
+        }
+        global $wpdb;
+        $event_ids = array_map(
+            'intval',
+            $event_ids
+        );
+        $imploded_sanitized_event_ids = implode(',', $event_ids);
+        // select transactions for events
+        // with subquery looking for other events
+        return array_map(
+            'intval',
+            $wpdb->get_col(
+                "SELECT 
+                      DISTINCT t.TXN_ID
+                    FROM 
+                      {$wpdb->prefix}esp_transaction t INNER JOIN 
+                      {$wpdb->prefix}esp_registration r ON t.TXN_ID=r.TXN_ID
+                    WHERE
+                       r.EVT_ID IN ({$imploded_sanitized_event_ids})
+                       AND NOT EXISTS 
+                       (
+                        SELECT 
+                          t.TXN_ID
+                        FROM 
+                          {$wpdb->prefix}esp_transaction ti INNER JOIN 
+                          {$wpdb->prefix}esp_registration ri ON ti.TXN_ID=ri.TXN_ID
+                        WHERE
+                          ti.TXN_ID=t.TXN_ID AND
+                          ri.EVT_ID NOT IN ({$imploded_sanitized_event_ids})
+                       )"
+            )
+        );
```

Comment:
not grokking this query at all...

it looks like you are running a near identical copy of the query as a subquery, then excluding the results of the inner from the outer results... ?!?!?!

> get TXN IDs where the Registration Event ID is in this list and not in this list ( get TXN IDs for the same TXN in the outer query where the Registration Event ID is in that same list )

so if the list of event IDs was `[ 1, 2, 3 ]` then it would be:

> get TXN IDs where the Registration Event ID is in [ 1, 2, 3 ] and not in ( get TXN IDs for the same TXN in the outer query where the Registration Event ID is in [ 1, 2, 3 ] )

it looks like the inner query will completely nullify the outer query since it is going to return a list of the same IDs that the outer query would find...

what am i missing?!?!


as well... the inner query does this:

```sql
SELECT t.TXN_ID ... ... WHERE ti.TXN_ID=t.TXN_ID AND ...
```

which looks to me like it is just going to return the TXN ID for whatever row the outer query happens to be looping over...

I'm soooo confused....

Reply:
no worries, ya that `NOT` is obviously easy to miss. It could use another comment I think...